File: android/src/main/kotlin/com/cloudwebrtc/webrtc/proxy/PeerObserver.kt
package com.cloudwebrtc.webrtc.proxy

import com.cloudwebrtc.webrtc.model.IceCandidate
import com.cloudwebrtc.webrtc.model.IceConnectionState
import com.cloudwebrtc.webrtc.model.IceGatheringState
import com.cloudwebrtc.webrtc.model.SignalingState
import org.webrtc.DataChannel
import org.webrtc.MediaStream
import org.webrtc.PeerConnection
import org.webrtc.RtpReceiver
import org.webrtc.IceCandidate as WIceCandidate

/**
 * Implementor of the [PeerConnection.Observer] which notifies [PeerConnectionProxy] about
 * [PeerConnection] events.
 *
 */
class PeerObserver : PeerConnection.Observer {
    /**
     * [PeerConnectionProxy] which will be notified about all events.
     */
    private var peer: PeerConnectionProxy? = null

    override fun onSignalingChange(signallingState: PeerConnection.SignalingState?) {
        if (signallingState != null) {
            peer?.observableEventBroadcaster()
                ?.onSignalingStateChange(
                    SignalingState.fromWebRtc(
                        signallingState
                    )
                )

        }
    }

    override fun onIceConnectionChange(iceConnectionState: PeerConnection.IceConnectionState?) {
        if (iceConnectionState != null) {
            peer?.observableEventBroadcaster()
                ?.onIceConnectionStateChange(
                    IceConnectionState.fromWebRtc(
                        iceConnectionState
                    )
                )
        }
    }

    override fun onIceGatheringChange(iceGatheringState: PeerConnection.IceGatheringState?) {
        if (iceGatheringState != null) {
            peer?.observableEventBroadcaster()
                ?.onIceGatheringStateChange(
                    IceGatheringState.fromWebRtc(
                        iceGatheringState
                    )
                )
        }
    }

    override fun onIceCandidate(candidate: WIceCandidate?) {
        if (candidate != null) {
            peer?.observableEventBroadcaster()
                ?.onIceCandidate(IceCandidate.fromWebRtc(candidate))
        }
    }

    // TODO(#34): we should prefer onTrack
    override fun onAddTrack(
        receiver: RtpReceiver?,
        mediaStreams: Array<out MediaStream>?
    ) {
        if (receiver != null) {
            val track = receiver.track()
            if (track != null) {
                val transceivers = peer?.getTransceivers()!!
                for (trans in transceivers) {
                    if (trans.getReceiver().id() == receiver.id()) {
                        peer?.observableEventBroadcaster()
                            ?.onTrack(MediaStreamTrackProxy(track), trans)
                    }
                }
            }
        }
        super.onAddTrack(receiver, mediaStreams)
    }

    override fun onRenegotiationNeeded() {
        peer?.observableEventBroadcaster()?.onNegotiationNeeded()
    }

    override fun onIceConnectionReceivingChange(receiving: Boolean) {}
    override fun onIceCandidatesRemoved(candidates: Array<out WIceCandidate>?) {}
    override fun onAddStream(stream: MediaStream?) {}
    override fun onRemoveStream(stream: MediaStream?) {}
    override fun onDataChannel(chan: DataChannel?) {}

    /**
     * Sets [PeerConnectionProxy] which will be notified about all events.
     */
    fun setPeerConnection(newPeer: PeerConnectionProxy) {
        peer = newPeer
    }
}

File: ios/Classes/proxy/PeerObserver.swift
import Dispatch
import WebRTC

/// Observer for a native `RTCPeerConnectionDelegate`.
class PeerObserver: NSObject, RTCPeerConnectionDelegate {
  /// `PeerConnectionProxy` into which callbacks will be provided.
  var peer: PeerConnectionProxy?

  override init() {}

  /// Sets the underlying `PeerConnectionProxy` for this `PeerObserver`.
  func setPeer(peer: PeerConnectionProxy) {
    self.peer = peer
  }

  /// Fires an `onSignalingStateChange` callback in the `PeerConnectionProxy`.
  func peerConnection(
    _ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState
  ) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onSignalingStateChange(
        state: SignalingState.fromWebRtc(state: stateChanged))
    }
  }

  /// Fires an `onIceConnectionStateChange` callback in the
  /// `PeerConnectionProxy`.
  func peerConnection(
    _ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState
  ) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onIceConnectionStateChange(
        state: IceConnectionState.fromWebRtc(state: newState))
    }
  }

  /// Fires an `onConnectionStateChange` callback in the `PeerConnectionProxy`.
  func peerConnection(
    _ peerConnection: RTCPeerConnection, didChange newState: RTCPeerConnectionState
  ) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onConnectionStateChange(
        state: PeerConnectionState.fromWebRtc(state: newState))
    }
  }

  /// Fires an `onIceGatheringStateChange` callback in the
  /// `PeerConnectionProxy`.
  func peerConnection(
    _ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState
  ) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onIceGatheringStateChange(
        state: IceGatheringState.fromWebRtc(state: newState))
    }
  }

  /// Fires an `onIceCandidate` callback in the `PeerConnectionProxy`.
  func peerConnection(
    _ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate
  ) {
    DispatchQueue.main.async {
      self.peer!.broadcastEventObserver().onIceCandidate(
        candidate: IceCandidate(candidate: candidate))
    }
  }

  /// Fires an `onTrack` callback in the `PeerConnectionProxy`.
  func peerConnection(
    _ peerConnection: RTCPeerConnection, didStartReceivingOn transceiver: RTCRtpTransceiver
  ) {
    DispatchQueue.main.async {
      let track = transceiver.receiver.track!
      self.peer!.broadcastEventObserver().onTrack(
        track: MediaStreamTrackProxy(track: track, deviceId: nil, source: nil),
        transceiver: RtpTransceiverProxy(transceiver: transceiver))
    }
  }

  /// Does nothing.
  func peerConnection(
    _ peerConnection: RTCPeerConnection, didAdd receiver: RTCRtpReceiver,
    streams mediaStreams: [RTCMediaStream]
  ) {
  }

  /// Does nothing.
  func peerConnection(
    _ peerConnection: RTCPeerConnection, didRemove receiver: RTCRtpReceiver
  ) {
    DispatchQueue.main.async {
      self.peer!.receiverRemoved(endedReceiver: receiver)
    }
  }

  /// Does nothing.
  func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
  }

  /// Does nothing.
  func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {
  }

  /// Does nothing.
  func peerConnection(_ peerConnection: RTCPeerConnection, didOpen stream: RTCDataChannel) {
  }

  /// Does nothing.
  func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {
  }

  /// Does nothing.
  func peerConnection(
    _ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]
  ) {
  }
}
